File: Commands/Lookup.cs
﻿using System.Diagnostics.CodeAnalysis;
using System.Text.RegularExpressions;
using adramelech.Common;
using adramelech.Configuration;
using adramelech.Extensions;
using adramelech.Utilities;
using Discord;
using Discord.Interactions;
using Discord.WebSocket;
using Flurl;
using Newtonsoft.Json.Serialization;

namespace Adramelech.Commands;

public class Lookup : InteractionModuleBase<SocketInteractionContext<SocketSlashCommand>>
{
    [SlashCommand("lookup", "Lookup a ip or domain")]
    public async Task LookupAsync([Summary("target", "The target to lookup")] string target)
    {
        await DeferAsync();

        var ip = LookupHelper.IpRegex().Match(target).Success
            ? target
            : await GetIpFromDomain(target);
        if (ip.IsFailure)
        {
            await Context.SendError("Failed to get ip from domain", true);
            return;
        }

        var response = await $"https://ipwho.is/{ip.Value}".Request<Whois>("curl", new SnakeCaseNamingStrategy());
        if (response.IsDefault() || !response.Success)
        {
            await Context.SendError("Failed to get whois information", true);
            return;
        }

        var mainField = $"""
                         **IP:** {ip.Value}
                         **Domain:** {(target != ip.Value ? target : "N/A")}
                         **Type:** {response.Type}
                         """;
        var locationField = $"""
                             **Continent:** {response.Continent}
                             **Country:** {response.Country} :flag_{response.CountryCode.ToLower()}:
                             **Region:** {response.Region}
                             **City:** {response.City}
                             **Latitude:** {response.Latitude}
                             **Longitude:** {response.Longitude}
                             **Postal:** {response.Postal}
                             """;
        var connectionField = $"""
                               **ASN:** {response.Connection.Asn}
                               **Organization:** {response.Connection.Org}
                               **ISP:** {response.Connection.Isp}
                               **Domain:** {response.Connection.Domain}
                               """;
        var timezoneField = $"""
                             **ID:** {response.Timezone.Id}
                             **Offset:** {response.Timezone.Offset}
                             **UTC:** {response.Timezone.Utc}
                             """;

        var mapsUrl = new Url("https://www.google.com")
            .AppendPathSegments("maps", "search", "/")
            .SetQueryParam("api", 1)
            .SetQueryParam("query", $"{response.Latitude},{response.Longitude}");

        await FollowupAsync(
            embed: new EmbedBuilder()
                .WithColor(BotConfig.EmbedColor)
                .WithTitle("Lookup")
                .WithDescription("For best results, search by ip")
                .AddField(":zap: Main", mainField)
                .AddField(":earth_americas: Location", locationField)
                .AddField(":satellite: Connection", connectionField)
                .AddField(":clock1: Timezone", timezoneField)
                .WithFooter("Powered by ipwho.is and da.gd")
                .Build(),
            components: new ComponentBuilder()
                .WithButton("Open in Google Maps", style: ButtonStyle.Link, url: mapsUrl.ToString(),
                    emote: new Emoji("🌎"))
                .Build());
    }

    private static async Task<Result<string>> GetIpFromDomain(string domain)
    {
        var response = await $"https://da.gd/host/{domain}".Request<string>();
        if (response.IsNullOrEmpty()) return new Exception("Failed to get ip from domain");

        response = response!.Trim();

        if (response.StartsWith("No"))
            return new Exception("No ip found for domain");

        return response.Contains(',') ? response[..response.IndexOf(',')] : response;
    }

    [SuppressMessage("ReSharper", "UnusedAutoPropertyAccessor.Local")]
    private struct Whois
    {
        public bool Success { get; set; }
        public string Type { get; set; }
        public string Continent { get; set; }
        public string Country { get; set; }
        public string CountryCode { get; set; }
        public string Region { get; set; }
        public string City { get; set; }
        public string Latitude { get; set; }
        public string Longitude { get; set; }
        public string Postal { get; set; }
        public SConnection Connection { get; set; }
        public STimezone Timezone { get; set; }

        internal struct SConnection
        {
            public int Asn { get; set; }
            public string Org { get; set; }
            public string Isp { get; set; }
            public string Domain { get; set; }
        }

        internal struct STimezone
        {
            public string Id { get; set; }
            public int Offset { get; set; }
            public string Utc { get; set; }
        }
    }
}

internal partial class LookupHelper
{
    [GeneratedRegex(@"\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b")]
    public static partial Regex IpRegex();
}

File: src/bot/commands/slash/lookup.ts
import { stripIndents } from 'common-tags';
import {
  ButtonStyle,
  ChatInputCommandInteraction,
  ComponentType,
  MessageFlags,
  SlashCommandBuilder,
} from 'discord.js';
import ky from 'ky';
import { z } from 'zod';
import type { Result } from '~/common/result';
import env from '#env';
import type { Command } from '#bot/types/command';
import { sendError } from '#bot/utils/sendError';

const schema = z.object({
  success: z.boolean(),
  type: z.union([z.literal('IPv4'), z.literal('IPv6')]),
  continent: z.string(),
  country: z.string(),
  country_code: z.string().length(2),
  region: z.string(),
  city: z.string(),
  latitude: z.number(),
  longitude: z.number(),
  postal: z.string(),
  connection: z.object({
    asn: z.number(),
    org: z.string(),
    isp: z.string(),
    domain: z.string(),
  }),
  timezone: z.object({
    id: z.string(),
    offset: z.number(),
    utc: z.string(),
  }),
});

export const command = <Command>{
  data: new SlashCommandBuilder()
    .setName('lookup')
    .setDescription('Lookup a domain or IP address')
    .addStringOption((option) =>
      option
        .setName('target')
        .setDescription('The domain or IP address to lookup')
        .setRequired(true),
    ),
  cooldown: true,
  uses: ['ipwhois.io', 'da.gd'],
  async execute(intr: ChatInputCommandInteraction) {
    await intr.deferReply();

    const target = intr.options.getString('target', true);
    const match = z.string().ip().safeParse(target);
    const ip: Result<string> = match.success
      ? { data: target }
      : await getIpFromDomain(target);

    if (ip.error) return await sendError(intr, 'Failed to get IP from domain');

    const response = await ky
      .get(`https://ipwho.is/${ip.data}`, {
        headers: {
          'User-Agent': 'curl', // :3
        },
      })
      .json();
    const { data, error } = schema.safeParse(response);
    if (error) return await sendError(intr, 'Failed to parse response');

    const mapsUrl = new URL('https://www.google.com/maps/search/');
    mapsUrl.searchParams.append('api', '1');
    mapsUrl.searchParams.append('query', `${data.latitude},${data.longitude}`);

    await intr.followUp({
      flags: MessageFlags.IsComponentsV2,
      components: [
        {
          type: ComponentType.Container,
          accent_color: env.EMBED_COLOR,
          components: [
            {
              type: ComponentType.TextDisplay,
              content: stripIndents`
              # Lookup
              **IP:** ${ip.data}
              **Domain:** ${target === ip.data ? 'None' : target}
              **Type:** ${data.type}
              `,
            },
            { type: ComponentType.Separator },
            {
              type: ComponentType.TextDisplay,
              content: stripIndents`
              ## Location
              **Continent:** ${data.continent}
              **Country:** ${data.country} :flag_${data.country_code.toLowerCase()}:
              **Region:** ${data.region}
              **City:** ${data.city}
              **Latitude:** ${data.latitude}
              **Longitude:** ${data.longitude}
              **Postal Code:** ${data.postal}
              `,
            },
            {
              type: ComponentType.ActionRow,
              components: [
                {
                  type: ComponentType.Button,
                  style: ButtonStyle.Link,
                  label: 'Open in Google Maps',
                  url: mapsUrl.toString(),
                  emoji: { name: '🌎' },
                },
              ],
            },
            { type: ComponentType.Separator },
            {
              type: ComponentType.TextDisplay,
              content: stripIndents`
              ## Connection
              **ASN:** ${data.connection.asn}
              **Organization:** ${data.connection.org}
              **ISP:** ${data.connection.isp}
              **Domain:** ${data.connection.domain}
              `,
            },
            { type: ComponentType.Separator },
            {
              type: ComponentType.TextDisplay,
              content: stripIndents`
              ## Timezone
              **ID:** ${data.timezone.id}
              **Offset:** ${data.timezone.offset}
              **UTC:** ${data.timezone.utc}
              `,
            },
            {
              type: ComponentType.TextDisplay,
              content: '> Powered by ipwhois.io and da.gd',
            },
          ],
        },
      ],
    });
  },
};

async function getIpFromDomain(domain: string): Promise<Result<string>> {
  const response = (await ky.get(`https://da.gd/host/${domain}`).text()).trim();
  if (!response || response.startsWith('No'))
    return { error: new Error('Failed to get IP from domain') };

  return {
    data: response.includes(',')
      ? response.substring(0, response.indexOf(','))
      : response,
  };
}
